### PR TITLE
Array support and fixes

### PR DESCRIPTION
**Pull request by thehabes**
http://lived-religion.rerum.io/deer-lr/test_loading.html

http://lived-religion.rerum.io/deer-lr/test_saving.html

You can test this version of DEER on these pages.  Forms and annotations are already set up to test the logic.  If you have questions around something, go to those pages and add breakpoints to see what is happening.  You can also use them as templates to make your own testing page(s).

### CODE REVIEW

File: js/deer-config.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ export default {
     TYPE: "deer-type",              // attribute, JSON-LD @type
     TEMPLATE: "deer-template",      // attribute, enum for custom template
     KEY: "deer-key",                // attribute, key to use for annotation
+    KEYDUPLICATE: "deer-key-duplicate", // attribute, notes input is a duplicate and app should ignore
```
[the remark this answers is not in the document]
Comment (thehabes) [reply js/deer-config.js]: Yes.  This is the value DEER will write to inputs that it detects are duplicates, pretty much just as a warning for the dev that it is going to be ignored.  It doesn't have to be a config, or even necessarily written (as you said, in these situations we could just change the DEER.KEY attribute so it will be ignored).  This just seemed to me to be the most obvious and hand-holdy.

File: js/deer-record.js [hunk 2]
```diff
@@ -69,154 +69,263 @@ export default class DeerReport {
...
+                .then((function(obj) {
                     try {
-                        for(let el of Array.from(this.inputs)) {
-                            if(el.getAttribute(DEER.KEY)===key){
-                                let assertedValue = UTILS.getValue(obj[key])
-                                if(Array.isArray(assertedValue)) {
-                                    for (const v of assertedValue) {
-                                        if(!el.value && (["string","number"].indexOf(typeof v)!==-1)){
-                                            el.value = v
+                        let inputElems = Array.from(this.inputs)
```
[the remark this answers is not in the document]
Comment (thehabes) [reply js/deer-record.js]: It was intentional.  It is because form.input is necessarily the primary loop and the guaranteed worst case O(n) operation because you have to look at them all not matter what (not true for `obj.keys`).  I wanted to optimize the overall processing/runtime by optimizing O(n) * O(some consequential function because of something you found about input n).

The only need to do an `obj.key` lookup is if an `input[deer-key='avalue']` exists.  It is also much more likely that an `obj.key` lookup will be faster than shallow copying and looping over a set of input HTMLElements and checking against an attribute, especially since working with HTMLElements is notoriously costly.  I believe you side step a bunch of this cost by storing the elements in `this.inputs` ahead of time, which means you aren't repeating DOM look ups (smart), but still.  

`obj.key` lookup is (worst case) O(n) just like looping over each `obj.key` is O(n), but looping over each `obj.key` and then checking each `input` for something guarantees the worst case of O(n) * O(m) where `n` and `m` are `obj.keys.length` and `form.inputs.length`, respectively.  If instead we loop over each input and check that it needs an `obj.key` lookup, we could end up with the best case scenario of O(0) * O(m), which avoids risking the exponential potential (in the case where m >= n) when we don't need it.

Not only that `Array.from(this.inputs).forEach(inpt=>inpt.addEventListener('input', () => inpt.$isDirty = true))` was called before this logic, which mean `O(n) + (O(n) * O(m)` which I felt we could collapse if the priority loop was to loop the `<input>`s then add the event listener when we find it has a DEER.KEY during that loop.  

Arrays.from() in general is also not free, since it makes a shallow copy of an iterable, which also happened for each `obj.key`, which felt like extra work.

It also made the test and error/warning feedback make more sense to me.  Detecting and handling duplicates, for example, was difficult doing it the other way.

In summary,  I feel that doing it this way cuts back on some extraneous processing and makes detect-and-handle of peculiarities smoother (since peculiarities are going to be caused by devs doing something funky with `<inputs>`).   I am open to other opinions, as I do not dare call this fact without proper scientific proof.

File: js/deer-record.js [hunk 3]
```diff
@@ -69,154 +69,263 @@ export default class DeerReport {
...
+                        for (let el of inputElems) {
+                            let deerKeyValue = (el.hasAttribute(DEER.KEY)) ? el.getAttribute(DEER.KEY) : ""
+                            let mapsToAnno = false
+                            if (deerKeyValue) {
+                                //Then this is a DEER form input.
+                                el.addEventListener('input', (e) => e.target.$isDirty = true)
+                                let assertedValue = ""
+                                if (flatKeys[deerKeyValue] !== i) {
+                                    UTILS.warning("Duplicate input " + DEER.KEY + " attribute value '" + deerKeyValue + "' detected in form.  This input will be ignored upon form submission and only the first instance will be respected.  See duplicate below.", el)
+                                    el.setAttribute(DEER.KEYDUPLICATE, "true")
```
[the remark this answers is not in the document]
Comment (thehabes) [reply js/deer-record.js]: This is a fringe I did not really think about.  In my mind, neither of these would have a DEER.KEY.  I would have made it so that a change to either of them would write to a hidden `<input>` that has DEER.KEY="priority" so that the annotation could not be confused or duplicated.  DEER can only know to make an annotation if DEER.KEY exists, and if DEER.KEY is repeated how is it supposed to know whether to duplicate or not?  What could the dev do to make it understand that it isn't supposed to make this annotation twice?

I think this is one of those "How DEER should be used" moments that we will keep coming across.  I am open to ideas for how to handle more complex forms like this.

File: js/deer-config.js [hunk 4]
```diff
@@ -14,8 +15,11 @@ export default {
     ITEMTYPE: "deer-item-type",     //attribute, specialty forms ('entity' by default)
     SOURCE: "deer-source",          // attribute, URI for asserting annotation
     EVIDENCE: "nv-evidence",        // attribute, URI for supporting evidence
+    ARRAYDELIMETER: "deer-array-delimeter", //attribute, denotes delimeter to use for array.join()
+    ARRAYTYPE : "deer-array-type", //attribute, defines whether this array is a list or a set
     
     INPUTS: ["input","textarea","dataset","select"], // array of selectors, identifies inputs with .value
+    CONTAINERS: ["ItemList","ItemListElement", "List", "Set", "list","set", "@list", "@set"], // array of supported list and set types the app will dig into for array values
```
[the remark this answers is not in the document]
Comment (thehabes) [reply js/deer-config.js]: oof, perhaps.  Let me know.
